fastvideo/v1/pipelines/implementations/hunyuan/hunyuan_pipeline.py

```diff
@@ -0,0 +1,190 @@
+        assert encoder_2 is not None, "Text encoder 2 is not found"
+        encoder_2.to(inference_args.device)
+        encoder_2.to(dtype=PRECISION_TO_TYPE[inference_args.text_encoder_precision])
+        encoder_2.requires_grad_(False)
+
+        tokenizer_2 = modules.pop("tokenizer_2")
+        assert tokenizer_2 is not None, "Tokenizer 2 is not found"
+
+        text_encoder_2 = TextEncoder(
+            text_encoder=encoder_2,
+            tokenizer=tokenizer_2,
+            # text_encoder_type="text_encoder_2",
+            max_length=inference_args.text_len_2,
+            # text_encoder_precision=inference_args.text_encoder_precision,
+            device=inference_args.device if not inference_args.use_cpu_offload else "cpu",
+        )
+        modules["text_encoder"] = text_encoder
+        modules["text_encoder_2"] = text_encoder_2
+
+
+    def setup_pipeline(self, inference_args: InferenceArgs):
+        self.add_stage("input_validation_stage", 
+                        InputValidationStage())
+        self.add_stage("prompt_encoding_stage_primary", 
+                       PromptEncodingStage(is_secondary=False))
+        self.add_stage("prompt_encoding_stage_secondary", 
+                       PromptEncodingStage(is_secondary=True))
+        self.add_stage("conditioning_stage", 
+                       ConditioningStage())
+        self.add_stage("timestep_preparation_stage", 
+                       TimestepPreparationStage())
+        self.add_stage("latent_preparation_stage", 
+                       LatentPreparationStage())
+        self.add_stage("denoising_stage", 
+                       DenoisingStage())
+        self.add_stage("decoding_stage", 
+                       DecodingStage())
+    
+    def initialize_pipeline(self, inference_args: InferenceArgs):
+        assert len(self._stages) > 0, "Pipeline stages are not set"
+        assert len(self._modules) > 0, "Pipeline modules are not set"
+
+
+        vae_scale_factor = 2**(len(self.vae.block_out_channels) - 1)
+        inference_args.vae_scale_factor = vae_scale_factor
+
+        self.image_processor = VaeImageProcessor(vae_scale_factor=vae_scale_factor)
+        self.register_modules({"image_processor": self.image_processor})
+
+
+        num_channels_latents = self.transformer.in_channels
+        inference_args.num_channels_latents = num_channels_latents
+
+
+    # TODO
+    def adjust_video_length(self, batch: ForwardBatch, inference_args: InferenceArgs):
+        """Adjust video length based on VAE version"""
+        video_length = batch.num_frames
+        batch.num_frames = (video_length - 1) // 4 + 1
+        return batch
+
+    
+    @torch.no_grad()
+    def forward(self, batch: ForwardBatch, inference_args: InferenceArgs) -> ForwardBatch:
+        logger.info(f"Running pipeline stages: {self._stage_name_mapping.keys()}")
+        logger.info(f"Batch: {batch}")
+        # for stage in self._stages:
+            # batch = stage(batch, inference_args)
```

Comment:
Definitely prefer this programming model; otherwise it's meaningless to have stages.

If users want to custom logic between stages, you should ask them to implement a custom stage (so ofc the API of wrapping custom logic to be a pipeline stage has to be clean and simple), and add that stage to the pipeline; otherwise it's hard to apply pipeline-level optimizations because you may not have the fully picture.